File: cpp/剑指offer_33_二叉搜索树的后序遍历序列_medium.cpp
// 二叉搜索树；根大于所有左子树，小于所有右子树
// 后序遍历： [ 左子树 | 右子树 | 根节点 ]
// 递归法
class Solution {
public:
    bool verifyPostorder(vector<int>& postorder) {
        return subTree(postorder, 0, postorder.size()-1);
    }
    bool subTree(vector<int>& tree, int begin, int end)
    {
        if(begin>=end) return true;
        int p = begin;
        while(tree[p]<tree[end]) p++;
        int m = p;
        while(tree[p]>tree[end]) p++;
        return p==end && subTree(tree, begin, m-1) && subTree(tree, m, end-1);
    }
};

// 辅助单调栈 
// 后序遍历数组的倒序
class Solution {
public:
    bool verifyPostorder(vector<int>& postorder) {
        vector<int> stack;
        int root = INT_MAX;
        for(int i=postorder.size()-1; i>-1; i--)
        {
            if (postorder[i]>root) return false;
            while(!stack.empty() && postorder[i]<stack.back())
            {
                root = stack.back();
                stack.pop_back();
            }
            stack.push_back(postorder[i]);
        }
        return true;
    } 
};

File: python/剑指offer_33_二叉搜索树的后序遍历序列_medium.py
# 二叉搜索树；根大于所有左子树，小于所有右子树
# 后序遍历： [ 左子树 | 右子树 | 根节点 ]
# 递归法
class Solution(object):
    def verifyPostorder(self, postorder):
        """
        :type postorder: List[int]
        :rtype: bool
        """
        def subtree(begin, end):
            if begin >= end:
                return True
            p = begin
            while postorder[p] < postorder[end]: p += 1
            m = p
            while postorder[p] > postorder[end]: p += 1
            return p==end and subtree(begin, m-1) and subtree(m, end-1)
        return subtree(0, len(postorder)-1)

# 辅助单调栈
class Solution(object):
    def verifyPostorder(self, postorder):
        """
        :type postorder: List[int]
        :rtype: bool
        """
        stack, root = [], float('+inf')
        for i in range(len(postorder)-1, -1, -1):
            if postorder[i] > root: return False
            while stack and postorder[i] < stack[-1]:
                root = stack.pop()
            stack.append(postorder[i])
        return True